### PR TITLE
[Intel MKL] Enabling Quantized Conv ops in native format mode

### PR DESCRIPTION
**Pull request by mahmoud-abuzaina**

### CODE REVIEW

File: tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc [hunk 1]
```diff
@@ -69,29 +70,33 @@ class ConvMklToTF : public OpsTestBase {
...
+                                                : "_MklQuantizedConv2D")
+            .Input(FakeInput(DT_QUINT8))  // Input
+            .Input(FakeInput(DT_QINT8))   // Filter
+            .Input(FakeInput(DT_FLOAT))   // Min input
+            .Input(FakeInput(DT_FLOAT))   // Max input
+            .Input(FakeInput(DT_FLOAT))   // Min filter
+            .Input(FakeInput(DT_FLOAT))   // Max filter
+            .Attr("Tinput", DataTypeToEnum<quint8>::v())
+            .Attr("Tfilter", DataTypeToEnum<qint8>::v())
+            .Attr("T", DataTypeToEnum<quint8>::v())
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc:85]: Should this be in `if (!NativeFormatEnabled())` instead?

File: tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc [hunk 2]
```diff
@@ -69,29 +70,33 @@ class ConvMklToTF : public OpsTestBase {
...
+            .Attr("out_type", DataTypeToEnum<qint32>::v())
+            .Attr("strides", {1, stride, stride, 1})
+            .Attr("padding", "SAME")
+            .Attr("_kernel", "QuantizedMklOp");
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc:89]: Same here. Should this be in `if (!NativeFormatEnabled())` instead?

File: tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_perchannel_test.cc [hunk 3]
```diff
@@ -151,30 +155,35 @@ TEST_F(QuantizedConv2DPerchannelTest, Small) {
...
+
+    // Convert the output tensor in MKL to TF format.
+    ConvMklToTF conv_comp;
+    conv_comp.ConvertMKL2TF<qint32>(DT_QINT32, output, output_mkl_metadata,
+                                    output_quantized);
+  }
+
+  Tensor output_float = QuantizedTensorToFloat<qint32>(
+      NativeFormatEnabled() ? output : output_quantized, output_min,
+      output_max);
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_perchannel_test.cc:186]: This is used often. Let's refactor this to a utility function.

File: tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc [hunk 4]
```diff
@@ -296,29 +309,33 @@ TEST_F(QuantizedConv2DTest, SmallS8) {
...
+          .Input(FakeInput(DT_FLOAT))  // Max input
+          .Input(FakeInput(DT_FLOAT))  // Min filter
+          .Input(FakeInput(DT_FLOAT))  // Max filter
+          .Attr("Tinput", DataTypeToEnum<qint8>::v())
+          .Attr("Tfilter", DataTypeToEnum<qint8>::v())
+          .Attr("T", DataTypeToEnum<quint8>::v())
+          .Attr("padding", "VALID")
+          .Attr("out_type", DataTypeToEnum<qint32>::v())
+          .Attr("strides", {1, stride, stride, 1})
+          .Attr("_kernel", "QuantizedMklOp");
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc:328]: Should `T` and `_kernel` be in `if (!NativeFormatEnabled)` instead?

File: tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc [hunk 5]
```diff
@@ -436,43 +459,50 @@ TEST_F(QuantizedConv2DTest, Small32Bit) {
...
+          .Input(FakeInput(DT_FLOAT))   // Min filter
+          .Input(FakeInput(DT_FLOAT))   // Max filter
+          .Attr("Tinput", DataTypeToEnum<quint8>::v())
+          .Attr("Tfilter", DataTypeToEnum<qint8>::v())
+          .Attr("T", DataTypeToEnum<quint8>::v())
+          .Attr("out_type", DataTypeToEnum<qint32>::v())
+          .Attr("strides", {1, stride, stride, 1})
+          .Attr("padding", "SAME")
+          .Attr("padding_list", {0, 0, 1, 1, 1, 1, 0, 0})
+          .Attr("_kernel", "QuantizedMklOp");
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc:495]: Should `T` and `_kernel` be in `if (!NativeFormatEnabled())` instead?

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and I'm so sorry for the delay!

File: tensorflow/core/kernels/mkl/mkl_conv_ops.cc [hunk 6]
```diff
@@ -1983,21 +1996,38 @@ REGISTER_KERNEL_BUILDER(Name("QuantizedConv2D")
...
+                                             qint32, false, false, true>);
 
 REGISTER_KERNEL_BUILDER(Name("QuantizedConv2DAndRequantize")
                             .Device(DEVICE_CPU)
                             .TypeConstraint<quint8>("Tinput")
                             .TypeConstraint<qint8>("Tfilter")
                             .TypeConstraint<qint8>("out_type"),
-                        NoOp);
+                        MklQuantizedConv2DOp<CPUDevice, quint8, qint32, qint8,
+                                             qint8, false, false, true>);
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_conv_ops.cc]: Probably was done by mistake. I changed it to float to be consistent. Thanks!

File: tensorflow/core/kernels/mkl/mkl_conv_ops.cc [hunk 7]
```diff
@@ -2124,21 +2176,23 @@ REGISTER_KERNEL_BUILDER(
...
-                            .TypeConstraint<qint8>("Tfilter")
-                            .TypeConstraint<quint8>("out_type"),
-                        NoOp);
+REGISTER_KERNEL_BUILDER(
+    Name("QuantizedConv2DAndReluAndRequantize")
+        .Device(DEVICE_CPU)
+        .TypeConstraint<quint8>("Tinput")
+        .TypeConstraint<qint8>("Tfilter")
+        .TypeConstraint<quint8>("out_type"),
+    MklQuantizedConv2DReluOp<CPUDevice, quint8, qint32, quint8, quint8, false,
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_conv_ops.cc]: Probably was done by mistake. I changed it to float to be consistent. Thanks!

File: tensorflow/core/kernels/mkl/mkl_conv_ops.cc [hunk 8]
```diff
@@ -2035,36 +2065,58 @@ REGISTER_KERNEL_BUILDER(Name("_MklQuantizedConv2DAndRequantize")
...
+                                             qint8, true, false, true>);
+
+REGISTER_KERNEL_BUILDER(Name("QuantizedConv2DWithBiasAndRequantize")
+                            .Device(DEVICE_CPU)
+                            .TypeConstraint<qint8>("Tinput")
+                            .TypeConstraint<qint8>("Tfilter")
+                            .TypeConstraint<float>("Tbias")
+                            .TypeConstraint<qint8>("out_type"),
+                        MklQuantizedConv2DOp<CPUDevice, qint8, float, qint8,
+                                             qint8, true, false, true>);
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_conv_ops.cc]: Thank you so much for taking the time to create this. I made the change for quantized ops in this file.

File: tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc (hunk 1 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc]: Yes, fixed.

File: tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc (hunk 2 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc:65]: no, per the MKL rewrite algorithm, label is required if the op name changes (i.e. adding _MklNative).

File: tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_perchannel_test.cc (hunk 3 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_perchannel_test.cc]: I added the refactored code to new file (mkl_kernel_util.h). I tried adding that code tensorflow/core/util/mkl_util.h but I had hard time trying to fix bazel circular dependency as this file depends on too many packages. So created this file where we can move common kernels-related code to.

File: tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc (hunk 4 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc]: For `T` yes. For `_kernel` no, per the MKL rewrite algorithm, label is required if the op name changes (i.e. adding _MklNative).

File: tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc (hunk 5 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc]: For `T` yes. For `_kernel` no, per the MKL rewrite algorithm, label is required if the op name changes (i.e. adding _MklNative).

**Comment by mahmoud-abuzaina** [comment]
Thank you for the valuable comments. I have addressed them.

File: tensorflow/core/kernels/mkl/mkl_kernel_util.h [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+  }
+  void TestBody() {}
+};
+
+// Helper function that converts tensor from MKL format to TF format.
+template <typename T>
+inline Tensor GetTFFormatTensor(DataType dtype,
+                                const Tensor& mkl_quantized_tensor,
+                                const Tensor* mkl_metadata_tensor) {
+  CHECK_NOTNULL(mkl_metadata_tensor);
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_kernel_util.h:57]: Nit: We don't use `CHECK` in new code. Please use `DCHECK` instead.
```suggestion
  DCHECK(mkl_metadata_tensor);
```

File: tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc [hunk 10]
```diff
@@ -178,12 +155,8 @@ class QuantizedConv2DTest : public OpsTestBase {
...
     if (!NativeFormatEnabled()) {
       const Tensor& output_mkl_metadata = *GetOutput(3);
-
-      ConvMklToTF conv_comp;
-      Tensor output_quantized;
-      conv_comp.ConvertMklToTF<qint32>(DT_QINT32, output, output_mkl_metadata,
-                                       output_quantized);
-
+      Tensor output_quantized =
+          GetTFFormatTensor<qint32>(DT_QINT32, output, &output_mkl_metadata);
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc]: Nit: Can we make `output_mkl_metadata` a `const Tensor*` instead? Same for lines 422, 510, 566, and 625.
It'd be great if we could rename all of them (including the ones that call `QuantizedToFloatTFFormat`) to `output_mkl_metadata_ptr`.

File: tensorflow/core/kernels/mkl/mkl_kernel_util.h [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    output = *GetOutput(0);
+  }
+  void TestBody() {}
+};
+
+// Helper function that converts tensor from MKL format to TF format.
+template <typename T>
+inline Tensor GetTFFormatTensor(DataType dtype,
+                                const Tensor& mkl_quantized_tensor,
+                                const Tensor* mkl_metadata_tensor) {
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_kernel_util.h:56]: Nit: Can we add `_ptr` at the end to differentiate it from normal tensors?
```suggestion
                                const Tensor* mkl_metadata_tensor_ptr) {
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the changes and for cleaning up the registrations even further! I have a few nits.

File: tensorflow/core/kernels/mkl/mkl_kernel_util.h [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+#define TENSORFLOW_CORE_KERNELS_MKL_MKL_KERNEL_UTIL_H_
+
+#ifdef INTEL_MKL
+
+#include "tensorflow/core/framework/node_def_builder.h"
+#include "tensorflow/core/kernels/ops_testutil.h"
+#include "tensorflow/core/kernels/quantization_utils.h"
+
+namespace tensorflow {
+
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_kernel_util.h]: Done.

File: tensorflow/core/kernels/mkl/mkl_kernel_util.h (hunk 9 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_kernel_util.h]: Done.

File: tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc (hunk 10 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_quantized_conv_ops_test.cc]: Sure. Done.

File: tensorflow/core/kernels/mkl/mkl_kernel_util.h (hunk 11 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_kernel_util.h]: Done.

**Comment by mahmoud-abuzaina** [comment]
I have addressed the remaining comments. Please have a look. Thank you!

**Review by penpornk — APPROVED** [review]
Thank you for the changes! This PR looks good to me now. I think I can just approve it and remove the `ready-to-pull` tag. (Will add it back after #45107 is merged.)

**Review by penpornk — APPROVED** [review]
Thank you for the PR and I'm sorry for the delay!

**Review by gbaned — APPROVED** [review]

**Pull request merged**